python/pyarrow/_dataset.pyx

```diff
@@ -2125,12 +2125,44 @@ cdef class ScanTask(_Weakrefable):
         -------
         record_batches : iterator of RecordBatch
         """
-        cdef shared_ptr[CRecordBatch] record_batch
+        # Return an explicit iterator object instead of using a
+        # generator so that this method is eagerly evaluated (a
+        # generator would mean no work gets done until the first
+        # iteration). This also works around a bug in Cython's
+        # generator.
+        cdef CRecordBatchIterator iterator
         with nogil:
-            for maybe_batch in GetResultValue(self.task.Execute()):
-                record_batch = GetResultValue(move(maybe_batch))
-                with gil:
-                    yield pyarrow_wrap_batch(record_batch)
+            iterator = move(GetResultValue(self.task.Execute()))
+            with gil:
+                return RecordBatchIterator.wrap(self, move(iterator))
+
+
+cdef class RecordBatchIterator(_Weakrefable):
+    """An iterator over a sequence of record batches."""
+    cdef:
+        ScanTask task
+        CRecordBatchIterator iterator
+
+    def __init__(self):
+        _forbid_instantiation(self.__class__, subclasses_instead=False)
+
+    @staticmethod
+    cdef wrap(ScanTask task, CRecordBatchIterator iterator):
+        cdef RecordBatchIterator self = \
+            RecordBatchIterator.__new__(RecordBatchIterator)
+        self.task = task
+        self.iterator = move(iterator)
+        return self
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        cdef shared_ptr[CRecordBatch] record_batch
+        record_batch = GetResultValue(move(self.iterator.Next()))
```

Comment:
You're right on both counts, I pushed a fixed version. (The `with gil` was because I was mimicking the original code but that doesn't make any sense anymore.)